### PR TITLE
feat: implement tda consent configuration support

### PR DESCRIPTION
**Pull request by Septimus4**
ENG-1394 implement tda consent configuration support

# Checklist

- [x] `make test` passes
- [x] tests and/or benchmarks are included
- [ ] documentation is changed or added
- [x] commit message follows [commit guidelines](https://github.com/indykite/.github/blob/master/commit-message.md)
- [ ] is breaking change

### CODE REVIEW

File: indykite/resource_consent.go [hunk 1]
```diff
@@ -0,0 +1,119 @@
+// Copyright (c) 2022 IndyKite
```
Comment (arxeiss) [inline indykite/resource_consent.go:1]: 2024

File: indykite/resource_consent.go [hunk 2]
```diff
@@ -0,0 +1,119 @@
...
+				Optional:    true,
+				Default:     false,
+				Description: "If set to true, the consent will be revoked after it is used once",
+			},
+			consentDataPointsKey: {
+				Type:        schema.TypeList,
+				Required:    true,
+				Elem:        &schema.Schema{Type: schema.TypeString},
+				Description: "Data points is a list of properties related to the Digital twin that the consent is for",
+			},
```
Comment (arxeiss) [inline indykite/resource_consent.go:81]: You are missing validations. Valid data are only in the description.

You can validate period key. `Application` can be checked by valid GID at least. There is also helper to validate Valid JSON.

Comment (arxeiss) [inline indykite/resource_consent.go:81]: I'm not sure this is good solution.

1. The order in array matters. I'm not sure it is intended. But if you just change order in the list, Terraform will be unhappy
2. If you format JSON differently, Terraform will be unhappy again. You should implement `DiffSuppressFunc: structure.SuppressJsonDiff,` for element.

File: indykite/resource_consent.go [hunk 3]
```diff
@@ -0,0 +1,119 @@
...
+
+func resourceConsentBuild(
+	_ *diag.Diagnostics,
+	data *schema.ResourceData,
+	_ *ClientContext,
+	builder *config.NodeRequest,
+) {
+	cfg := consentConfigBuilder(data)
+	builder.WithConsentConfig(cfg)
+}
```
Comment (arxeiss) [inline indykite/resource_consent.go:119]: Nitpick: This seems too superfluous to me. I don't see a need for another function.

**Review by arxeiss — CHANGES_REQUESTED** [review]

File: examples/resources/indykite_consent_configuration/resource.tf [hunk 4]
```diff
@@ -0,0 +1,13 @@
+resource "indykite_consent" "basic-user-data" {
```
Comment (arxeiss) [inline examples/resources/indykite_consent_configuration/resource.tf:1]: Wrong file location. Folder should be named same way the resource is

```diff
-examples/resources/indykite_consent_configuration/resource.tf
+examples/resources/indykite_consent/resource.tf
```

**Review by arxeiss — CHANGES_REQUESTED** [review]

File: indykite/resource_consent.go [hunk 5]
```diff
@@ -0,0 +1,132 @@
...
+				Required:     true,
+				Description:  "Purpose of the consent",
+				ValidateFunc: validation.StringIsNotEmpty,
+			},
+			consentApplicationIDKey: {
+				Type:             schema.TypeString,
+				Required:         true,
+				Description:      "Application ID",
+				ValidateDiagFunc: ValidateGID,
+			},
```
Comment (arxeiss) [inline indykite/resource_consent.go:70]: Can you use `applicationIDSchema()`?

File: indykite/resource_consent.go [hunk 6]
```diff
@@ -0,0 +1,132 @@
...
+					var errs []error
+					v := val.(int)
+					if v < 86400 || v > 63072000 {
+						errs = append(errs,
+							fmt.Errorf("%q must be between 86400 (1 day) and 63072000 (2 years) seconds, got: %d",
+								key,
+								v))
+					}
+					return nil, errs
+				},
```
Comment (arxeiss) [inline indykite/resource_consent.go:85]: Why you have custom validation function? Why not just use `validation.IntBetween(86400, 63072000)`?

File: indykite/resource_consent.go [hunk 7]
```diff
@@ -0,0 +1,132 @@
...
+				Default:     false,
+				Description: "If set to true, the consent will be revoked after it is used once",
+			},
+			consentDataPointsKey: {
+				Type:             schema.TypeList,
+				Required:         true,
+				Elem:             &schema.Schema{Type: schema.TypeString},
+				Description:      "Data points is a list of properties related to the Digital twin that the consent is for",
+				DiffSuppressFunc: structure.SuppressJsonDiff,
+			},
```
Comment (arxeiss) [inline indykite/resource_consent.go:99]: 1. Add validation, that elem is JSON
2. Write a test which will validate it

File: indykite/resource_consent_test.go [hunk 8]
```diff
@@ -0,0 +1,401 @@
...
+				DisplayName: "Wonka Consent for chocolate receipts",
+				CreateTime:  timestamppb.Now(),
+				UpdateTime:  timestamppb.Now(),
+				Config: &configpb.ConfigNode_ConsentConfig{
+					ConsentConfig: &configpb.ConsentConfiguration{
+						Purpose:        "To allow Wonka to store and process chocolate receipts",
+						ApplicationId:  applicationID,
+						RevokeAfterUse: true,
+						ValidityPeriod: 86400,
+						DataPoints:     []string{"{\"returns\": [{\"properties\": [\"name\", \"location\"]}]}"},
```
Comment (arxeiss) [inline indykite/resource_consent_test.go:93]: Suggestion: If you use ``` ` ``` you don't have to escape every `"`"

**Review by arxeiss — CHANGES_REQUESTED** [review]

File: indykite/resource_consent_test.go (hunk 8 above)
Comment (Septimus4) [reply indykite/resource_consent_test.go:93]: The use of single quotes is not allowed in terraform files

Comment (arxeiss) [reply indykite/resource_consent_test.go:93]: 1. This is Go code, not Terraform code. You can use ``` ` ``` here
2. In Terraform code you can use `<<-EOT` or 

```
jsonencode({
	"kid": "abc",
	"use": "sig",
	"alg": "RS256",
	"n": "--nothing-real-just-random-xyqwerasf--",
	"kty": "RSA"
}),
```

Comment (Septimus4) [reply indykite/resource_consent_test.go]: Ah yes sorry I  miss red the file, I  thought it was the example file, I will also fix the example file with jsonencode just to be readable

**Review by arxeiss — APPROVED** [review]
Have you tested that the tests are passing?
I mean the integration tests inside tests folder

**Comment by Septimus4** [comment]
I tested locally and you can also see the tests in https://github.com/indykite/terraform-provider-indykite/actions/runs/10109999277/job/27959020793?pr=74

**Pull request merged**